### PR TITLE
Handle ForeignKey String References

### PR DESCRIPTION
**Pull request by alejandro-angulo**
Possible solution to https://github.com/PyCQA/pylint-django/issues/243 . This change leverages Django's model name resolution to handle ForeignKey string references.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -95,9 +95,10 @@ This plugin is disabled by default! To enable it::
 Known issues
 ------------
 
-If you reference foreign-key models by their name (as string) ``pylint-django`` may not be
-able to find the model and will report issues because it has no idea what the underlying
-type of this field is. Supported options are::
+If Django is not installed and  you reference foreign-key models by their name (as
+string) ``pylint-django`` may not be able to find the model and will report issues
+because it has no idea what the underlying type of this field is. Supported options
+are::
```
Comment (atodorov) [inline README.rst:101]: Why the change here ? I don't see anything different in the text.

File: pylint_django/transforms/foreignkey.py [hunk 2]
```diff
@@ -87,10 +100,24 @@ def infer_key_classes(node, context=None):
...
+                use_django_model_resolution = bool(django_spec)
+
+                if use_django_model_resolution:
+                    # If Django is installed we can use it to resolve the module name
+                    try:
+                        module_name = _module_name_from_django_model_resolution(model_name, module_name)
+                    except LookupError:
+                        use_django_model_resolution = False
+
+                if not use_django_model_resolution:
```
Comment (atodorov) [inline pylint_django/transforms/foreignkey.py:115]: that's an `else` to the above block, isn't it ?

File: pylint_django/tests/test_app/apps.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from django.apps import AppConfig
+
+
+class TestAppConfig(AppConfig):
+    name = 'test_app'
```
Comment (atodorov) [inline pylint_django/tests/test_app/apps.py:5]: I believe all of this should be under `input/`. I can see a `models/author.py` module in there.

File: pylint_django/tests/test_func.py [hunk 4]
```diff
@@ -6,6 +6,8 @@
 import pylint
 
 
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'pylint_django.tests.conftest')
+
```
Comment (atodorov) [inline pylint_django/tests/test_func.py:10]: Why do we need this? checkers/django_installed.py also tries to import Django and it used to work without anything else so please explain.

File: pylint_django/transforms/foreignkey.py [hunk 5]
```diff
@@ -87,10 +100,24 @@ def infer_key_classes(node, context=None):
...
             elif not module_name.endswith('models'):
                 # otherwise Django allows specifying an app name first, e.g.
-                # ForeignKey('auth.User') so we try to convert that to
-                # 'auth.models', 'User' which works nicely with the `endswith()`
-                # comparison below
-                module_name += '.models'
+                # ForeignKey('auth.User')
+
+                django_spec = find_spec('django')
+                use_django_model_resolution = bool(django_spec)
```
Comment (atodorov) [inline pylint_django/transforms/foreignkey.py:106]: is this making sure that Django is installed ? If yes, then you don't need it b/c there's checkers/django_installed.py which will produce a fatal failure if you try executing pylint-django in environment which doesn't have Django.

File: pylint_django/transforms/foreignkey.py (hunk 2 above)
Comment (alejandro-angulo) [reply pylint_django/transforms/foreignkey.py:115]: Yes, but I didn't use an `else` because I wanted to run this logic if django's resolution failed. The value of `use_django_model_resolution` is updated on line 113, inside of the `if` block above.

File: pylint_django/tests/test_app/apps.py (hunk 3 above)
Comment (alejandro-angulo) [reply pylint_django/tests/test_app/apps.py:5]: Ah okay, I wasn't sure if this belonged in there since it's moreso setup than something to test. I'll move this into `input/`.

File: pylint_django/transforms/foreignkey.py (hunk 5 above)
Comment (alejandro-angulo) [reply pylint_django/transforms/foreignkey.py:106]: Yeah this is just a check to see if Django is installed. I didn't know about that check. I'll remove this.

File: README.rst (hunk 1 above)
Comment (alejandro-angulo) [reply README.rst:101]: I added some additional text to the first line `If Django is not installed...`. But this was before I knew about `checkers/django_installed.py` so I'll undo this change (don't think it makes much sense anymore).

File: pylint_django/tests/test_func.py (hunk 4 above)
Comment (alejandro-angulo) [reply pylint_django/tests/test_func.py:10]: I added this so that the test configuration would be used when `django.setup` is called in `_module_name_from_django_model_resolution` (https://github.com/PyCQA/pylint-django/pull/272/files#diff-1bbed731b242312a86f3f314c3682343R47). But looking back on it, I think it'd be better to call `settings.configure` before `django.setup` to avoid messing with environment variables.

File: README.rst (hunk 1 above)
Comment (alejandro-angulo) [reply README.rst]: I undid my earlier change but maybe the entire known issues section should be remove if this is accepted.

File: pylint_django/transforms/foreignkey.py (hunk 2 above)
Comment (alejandro-angulo) [reply pylint_django/transforms/foreignkey.py]: No longer checking to see if Django is installed so these specific ifs aren't needed anymore. Resolving conversation.

File: pylint_django/tests/test_func.py (hunk 4 above)
Comment (atodorov) [reply pylint_django/tests/test_func.py:10]: Makes sense then. Let's name this module settings.py though.

File: pylint_django/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import os
+
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```
Comment (atodorov) [inline pylint_django/tests/conftest.py:4]: This variable and the import above don't seem to be needed.

File: pylint_django/transforms/foreignkey.py [hunk 7]
```diff
@@ -41,6 +41,18 @@ def _get_model_class_defs_from_module(module, model_name, module_name):
     return class_defs
 
 
+def _module_name_from_django_model_resolution(model_name, module_name):
+    import django  # pylint: disable=import-outside-toplevel
```
Comment (atodorov) [inline pylint_django/transforms/foreignkey.py:45]: IMO this can be a top-level import, no ?

File: pylint_django/transforms/foreignkey.py [hunk 8]
```diff
@@ -41,6 +41,18 @@ def _get_model_class_defs_from_module(module, model_name, module_name):
...
+    django.setup()
+    from django.apps import apps  # pylint: disable=import-outside-toplevel
```
Comment (atodorov) [inline pylint_django/transforms/foreignkey.py:47]: same for this one, top-level ?

File: pylint_django/transforms/foreignkey.py [hunk 9]
```diff
@@ -41,6 +41,18 @@ def _get_model_class_defs_from_module(module, model_name, module_name):
...
+
+    app = apps.get_app_config(module_name)
+    model = app.get_model(model_name)
+    module_name = model.__module__
```
Comment (atodorov) [inline pylint_django/transforms/foreignkey.py:51]: can directly return this value instead of using an intermediate variable.

File: pylint_django/tests/input/func_noerror_string_foreignkey.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+"""
+Checks that PyLint correctly handles string foreign keys
+https://github.com/PyCQA/pylint-django/issues/243
+"""
+#  pylint: disable=missing-docstring
+from django.db import models
+
+
+class Book(models.Model):
+    author = models.ForeignKey("test_app.Author", models.CASCADE)
```
Comment (atodorov) [inline pylint_django/tests/input/func_noerror_string_foreignkey.py:10]: Let's also add another FK reference to "auth.User" - because django apps don't follow the usual convention of app name vs. module name. That's was one of the reasons we didn't deal with string FKs for so long.

File: pylint_django/tests/conftest.py (hunk 6 above)
Comment (alejandro-angulo) [reply pylint_django/tests/conftest.py:4]: Oh good catch. This was left over from before when I was checking to see if Django was installed. Will fix.

File: pylint_django/transforms/foreignkey.py (hunk 8 above)
Comment (alejandro-angulo) [reply pylint_django/transforms/foreignkey.py:47]: Oh good catch. This was left over from before when I was checking to see if Django was installed. Will fix.

File: pylint_django/tests/conftest.py (hunk 6 above)
Comment (alejandro-angulo) [reply pylint_django/tests/conftest.py:4]: Good point. Don't think `DEBUG` is needed either.

File: pylint_django/tests/settings.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+SECRET_KEY = 'fake-key'
+
+INSTALLED_APPS = [
+    'django.contrib.auth',
```
Comment (alejandro-angulo) [inline pylint_django/tests/settings.py:4]: These two apps and the two middleware are required for Django's authentication features (e.g. `auth.User`): https://docs.djangoproject.com/en/3.0/topics/auth/#installation .

File: pylint_django/transforms/foreignkey.py [hunk 12]
```diff
@@ -41,6 +41,17 @@ def _get_model_class_defs_from_module(module, model_name, module_name):
     return class_defs
 
 
+def _module_name_from_django_model_resolution(model_name, module_name):
+    import django  # pylint: disable=import-outside-toplevel
+    django.setup()
+    from django.apps import apps  # pylint: disable=import-outside-toplevel
```
Comment (alejandro-angulo) [inline pylint_django/transforms/foreignkey.py:47]: Moving the imports into the top-level broke the build during the `django_not_installed` check: https://travis-ci.org/github/PyCQA/pylint-django/builds/710245712 .

**Comment by ajhodges** [comment]
https://github.com/PyCQA/pylint-django#known-issues

I think the Readme can probably be updated now 👍👍

**Comment by 515hikaru** [comment]
@alejandro-angulo  I was just in trouble facing this problem. Thank you for fixing it!
